data-node/src/main/java/org/graylog/datanode/bootstrap/preflight/DataNodeCertRenewalPeriodical.java

```diff
@@ -46,23 +50,28 @@ public class DataNodeCertRenewalPeriodical extends Periodical {
 
     private final Supplier<Boolean> isServerInPreflightMode;
 
+    /**
+     * We need to differentiate between interval of the task scheduling and actual automatic CSR request triggering.
+     * This prevents flooding the server with CSR requests, which would then lead to repeated certificate signing and
+     * repeated opensearch process reboots. We want to check if the CSR is needed very often,
+     * but we don't want to trigger it many times in a row.
+     */
+    private final RateLimiter rateLimiter;
```

Comment:
I am not too worried about the Beta annotation. It's been there for years and years without any problems. https://github.com/google/guava/issues/5612

In the same time it feels like the cleanest and very simple solution that doesn't require any complex implementation on our side. Whatever we do will be more fragile, harder to read and error prone.

Reply:
Agreed and I see why you used it. However, it still leaves a bad taste.

There's a mismatch between the comment and the actual creation of the rate limiter by the way. Looks like it is configured to be only running once every 5 minutes.
Since the CSR process will never take longer than that, I though using a simple timestamp set when triggering it and check against that would be sufficient.